File: android/core/pitch/src/main/kotlin/com/vijaychhetry/kidspiano/core/pitch/Synth.kt
package com.vijaychhetry.kidspiano.core.pitch

fun pianoTone(
    frequency: Double,
    sampleRate: Int,
    durationSec: Double,
    amplitude: Float = 0.35f,
    harmonics: Int = 6,
): FloatArray {
    val n = (sampleRate * durationSec).toInt()
    val out = FloatArray(n)
    val weight = (1..harmonics).sumOf { 1.0 / it }
    for (i in 0 until n) {
        val t = i.toDouble() / sampleRate
        var sample = 0.0
        for (h in 1..harmonics) {
            val w = 1.0 / h
            sample += w * Math.sin(2 * Math.PI * frequency * h * t) * Math.exp(-2.2 * h * 0.15 * t)
        }
        val attack = minOf(1.0, t / 0.008)
        out[i] = (amplitude * attack * (sample / weight)).toFloat()
    }
    return out
}

/**
 * Closer to a miked piano than [pianoTone]: stretched (inharmonic) partials,
 * a hammer noise burst, and mains hum from the room.
 */
fun realisticPianoTone(
    frequency: Double,
    sampleRate: Int,
    durationSec: Double,
    amplitude: Float = 0.3f,
    partials: Int = 8,
    inharmonicity: Double = 0.0004,
    noiseLevel: Double = 0.02,
    humLevel: Double = 0.01,
    seed: Long = 42L,
): FloatArray {
    val n = (sampleRate * durationSec).toInt()
    val out = FloatArray(n)
    val random = java.util.Random(seed)
    val weight = (1..partials).sumOf { 1.0 / Math.pow(it.toDouble(), 1.2) }
    for (i in 0 until n) {
        val t = i.toDouble() / sampleRate
        var sample = 0.0
        for (h in 1..partials) {
            val stretched = h * frequency * Math.sqrt(1.0 + inharmonicity * h * h)
            if (stretched >= sampleRate / 2.0) continue
            val w = 1.0 / Math.pow(h.toDouble(), 1.2)
            sample += w * Math.sin(2 * Math.PI * stretched * t) * Math.exp(-0.9 * h * 0.18 * t)
        }
        sample /= weight
        val hammer = random.nextGaussian() * noiseLevel * Math.exp(-t / 0.02)
        val hum = humLevel * Math.sin(2 * Math.PI * 50.0 * t)
        val attack = minOf(1.0, t / 0.006)
        out[i] = (amplitude * attack * sample + hammer + hum).toFloat()
    }
    return out
}

fun mixTones(a: FloatArray, b: FloatArray): FloatArray {
    val n = minOf(a.size, b.size)
    val out = FloatArray(n)
    for (i in 0 until n) out[i] = a[i] + b[i]
    return out
}

fun sineTone(frequency: Double, sampleRate: Int, durationSec: Double, amplitude: Float = 0.4f): FloatArray {
    val n = (sampleRate * durationSec).toInt()
    val out = FloatArray(n)
    for (i in 0 until n) {
        out[i] = (amplitude * Math.sin(2 * Math.PI * frequency * i / sampleRate)).toFloat()
    }
    return out
}

File: src/audio/synth.ts
/** Synthetic piano-like tones for tests and the in-app simulator. */

export function pianoTone(
  frequency: number,
  sampleRate: number,
  durationSec: number,
  options: { amplitude?: number; harmonics?: number; decay?: number } = {},
): Float32Array {
  const amplitude = options.amplitude ?? 0.35;
  const harmonics = options.harmonics ?? 6;
  const decay = options.decay ?? 2.2;
  const n = Math.floor(sampleRate * durationSec);
  const out = new Float32Array(n);
  for (let i = 0; i < n; i++) {
    const t = i / sampleRate;
    let sample = 0;
    let weight = 0;
    for (let h = 1; h <= harmonics; h++) {
      const w = 1 / h;
      sample += w * Math.sin(2 * Math.PI * frequency * h * t) * Math.exp(-decay * h * 0.15 * t);
      weight += w;
    }
    const attack = Math.min(1, t / 0.008);
    out[i] = amplitude * attack * (sample / weight);
  }
  return out;
}

export function sineTone(frequency: number, sampleRate: number, durationSec: number, amplitude = 0.4): Float32Array {
  const n = Math.floor(sampleRate * durationSec);
  const out = new Float32Array(n);
  for (let i = 0; i < n; i++) {
    out[i] = amplitude * Math.sin((2 * Math.PI * frequency * i) / sampleRate);
  }
  return out;
}

export function mixNoise(signal: Float32Array, noiseRms = 0.002): Float32Array {
  const out = new Float32Array(signal.length);
  for (let i = 0; i < signal.length; i++) {
    const noise = (Math.random() * 2 - 1) * noiseRms * Math.sqrt(3);
    out[i] = signal[i] + noise;
  }
  return out;
}

export function sliceFrame(signal: Float32Array, start: number, length: number): Float32Array {
  return signal.slice(start, start + length);
}
